### ISSUE TITLE
Oracle. Cannot pass blob >=32 768 bytes as stored procedure parameter

### ISSUE THREAD

**Issue by IgorBuchelnikov** [linq2db/linq2db#1999]
```
DataParameter a_data = new DataParameter("a_data", new byte[32768], DataType.Blob);
a_data.Size = alert.alert.photo.content.Length;
dataConnection.ExecuteProc("ProcedureName", new[] {  a_data });
```

> ORA-01460: unimplemented or unreasonable conversion requested blob insert

Following code works:

```
var cmd = new OracleCommand("declare xx blob; begin dbms_lob.createtemporary(xx, false, 0); :tempblob := xx; end;");
cmd.Connection = (OracleConnection)dataConnection.Connection;

var param = new OracleParameter("tempblob", OracleDbType.Blob) { Direction = ParameterDirection.Output };
cmd.Parameters.Add(param);

cmd.ExecuteNonQuery();

var tempLob = (OracleBlob)param.Value;
tempLob.Write(new byte[32768], 0, 32768);

var comm = dataConnection.Connection.CreateCommand();
comm.CommandText = "ProcedureName";
comm.CommandType = CommandType.StoredProcedure;

var a_data = new OracleParameter("a_data", OracleDbType.Blob) { Value = tempLob };
comm.Parameters.Add(a_data );

comm.ExecuteNonQuery();
```

**Comment by MaceWindu** [comment]
Yep, our oracle support level needs improvements. Very 'special' db..

**Comment by MaceWindu** [comment]
Cannot test right now, but you probably can workaround it by registering proper parameter converter in mapping schema.

**Comment by IgorBuchelnikov** [comment]
If you direct me what files to change and how to create test I will create pull request.

**Comment by MaceWindu** [comment]
WIll try to check it in an hour

**Comment by MaceWindu** [comment]
Looks like Oracle have really strange limitations in this area that will require changes to our provider.
Unfortunatelly I don't have time this week to to dig into it. So if I will came with something, it will not be earlier than weekend.

**Comment by mrcopperbeard** [comment]
Found a workaround by specifying `dbType` as "BLOB" explicitly via constructor
```csharp
new DataParameter("a_data", tempLob, "BLOB")
```

Many thanks to [fellows from EF Core](https://github.com/dotnet/efcore/issues/14190) ;)

**Comment by jods4** [comment]
I noticed something similar recently:

Imagine you do this query:
```csharp
var blob = new OracleBlob();
// fill the blob with data, etc.
db.TargetTable.Insert(() => new TargetTable 
{
  Data = blob
});
```
It _seems_ to work (up to the Oracle size limit for binary) but if you look at Linq2Db logs, you'll notice that the whole blob data is listed as the parameter value.

So my guess is that _somehow_ Linq2Db managed to read back the Stream and passes it as a BINARY parameter rather than a native BLOB / OracleBlob.

@MaceWindu does Linq2Db read streams when they're used as parameters?
If so, a simple fix might be to whitelist per-provider native types and pass them unchanged to parameters?

**Comment by MaceWindu** [comment]
We don't support streams right now in any way. There are couple of issues regarding it already I think

**Comment by jods4** [comment]
@MaceWindu 
> We don't support streams right now in any way.

Ok, so the behavior when passing an `OracleBlob` value is just weird. "Something" must read it back and put it as a BINARY parameter.
It would be great if Linq2Db could recognize `OracleBlob` as a native DB type and pass it straight into the parameter value.

That would unlock a query where we had no choice but fallback to raw sql string.

**Comment by MaceWindu** [comment]
To pass it to parameter is should be enough to specify coversion to `DataParameter` I suppose. I'm more bothered by disposal policy:
- who and when should dispose streams/disposable parameters on write
- how ownership works on data read - should it be disposed by user or it will be disposed automatically after disposal of e.g. data reader (which also means we cannot dispose it automatically anymore and should have some new cleanup mechanisms)

Also streams supported not only by Oracle provider but by mssql too at least.

**Comment by jods4** [comment]
Yes, there is another Github issue about having a true, generic, stream support. Disposal is a big issue and differences between dbs as well: when writing Oracle works in push fashion (you write into `OracleBlob` first); SqlServer works in a pull fashion (you pass a stream that is read by the sql provider).

In the meantime the ability to work with native DB types (e.g. `OracleBlob`) is a good enough stop-gap.

I found out that if you type your model property as `OracleBlob` you can read without problems:
```csharp
class Frob
{
  public int Id { get; set }
  public OracleBlob Data { get; set; }
}

var row = db.Frobs.First(x => x.Id == 3);
// This is a perfectly usable OracleBlob that you can read as a stream.
var blob = row.Data;
// Don't forget to dispose.
```

But strangely something fishy happens if you try to write:
```csharp
var frob = new Frob { Id = 1, Data = new OracleBlob(...) };
// Write into the blob
frob.Data.Write(...);
// Try to insert
db.Frobs.Insert(frob);
// Works, but not as you'd expect....
```
This write _seems_ to work but as was observed in this issue, it also seems that the `OracleBlob` isn't passed directly as a parameter value but rather is converted into a byte array. For instance, every byte is logged in the trace (should not be possible) and the OP here observed that values larger than 32k don't go through (they should when using OracleBlob).

It's probably just a small issue but if it worked it would unlock writing large blobs with Linq2Db.

**Comment by MaceWindu** [comment]
Fixed long time ago in 3.4.0
